central/declarativeconfig/updater/notifier_updater.go

```diff
@@ -0,0 +1,108 @@
+	"github.com/stackrox/rox/pkg/notifier"
+	"github.com/stackrox/rox/pkg/notifiers"
+	"github.com/stackrox/rox/pkg/set"
+)
+
+type notifierUpdater struct {
+	notifierDS    notifierDataStore.DataStore
+	policyCleaner policycleaner.PolicyCleaner
+	processor     notifier.Processor
+	reporter      integrationhealth.Reporter
+	nameExtractor types.NameExtractor
+	idExtractor   types.IDExtractor
+}
+
+var _ ResourceUpdater = (*notifierUpdater)(nil)
+
+func newNotifierUpdater(notifierDS notifierDataStore.DataStore, policyCleaner policycleaner.PolicyCleaner, processor notifier.Processor, reporter integrationhealth.Reporter) ResourceUpdater {
+	return &notifierUpdater{
+		notifierDS:    notifierDS,
+		policyCleaner: policyCleaner,
+		processor:     processor,
+		reporter:      reporter,
+		idExtractor:   types.UniversalIDExtractor(),
+		nameExtractor: types.UniversalNameExtractor(),
+	}
+}
+
+func (u *notifierUpdater) Upsert(ctx context.Context, m proto.Message) error {
+	notifierProto, ok := m.(*storage.Notifier)
+	if !ok {
+		return errox.InvariantViolation.Newf("wrong type passed to role updater: %T", notifierProto)
+	}
+	_, err := u.notifierDS.UpsertNotifier(ctx, notifierProto)
+	if err != nil {
+		return err
+	}
+	notifier, err := notifiers.CreateNotifier(notifierProto)
+	if err != nil {
+		return errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+	u.processor.UpdateNotifier(ctx, notifier)
+
+	return u.reporter.Register(notifierProto.GetId(), notifierProto.GetName(), storage.IntegrationHealth_NOTIFIER)
+}
+
+func (u *notifierUpdater) DeleteResources(ctx context.Context, resourceIDsToSkip ...string) ([]string, error) {
+	notifiersToSkip := set.NewFrozenStringSet(resourceIDsToSkip...)
+
+	notifiers, err := u.notifierDS.GetNotifiersFiltered(ctx, func(n *storage.Notifier) bool {
+		return declarativeconfig.IsDeclarativeOrigin(n) &&
+			!notifiersToSkip.Contains(n.GetId())
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving declarative notifiers")
+	}
+
+	var notifierDeletionErr *multierror.Error
+	var notifierIDs []string
+	for _, n := range notifiers {
+		err = u.policyCleaner.DeleteNotifierFromPolicies(n.GetId())
+		if err != nil {
+			err := errors.Wrap(err, "deleting notifier from policies")
+			notifierDeletionErr, notifierIDs = u.processDeletionError(notifierDeletionErr, err, notifierIDs, n)
+			continue
+		}
+
+		// In case of temporary issues with database(for example, connectivity issues)
+		// it is possible that notifier is already deleted from datastore
+		// while integration health isn't.
+		if err := u.notifierDS.RemoveNotifier(ctx, n.GetId()); err != nil && !errors.Is(err, errox.NotFound) {
+			err := errors.Wrap(err, "deleting notifier from database")
+			notifierDeletionErr, notifierIDs = u.processDeletionError(notifierDeletionErr, err, notifierIDs, n)
+			continue
+		}
+
+		u.processor.RemoveNotifier(ctx, n.GetId())
+		if err := u.reporter.RemoveIntegrationHealth(n.GetId()); err != nil {
```

Comment:
Hmm, how will this work in case of notifiers, which are seemingly added "twice" under different types in the integration health reporter?

AFAICT, there will be right now one integration health entry created upon first creation of the notifier as declarative config health status, however once added within the processor and an alert is received, the integration health will be updated, using the same ID, this time however as a notifier integration health type.

While this won't break things immediately, it certainly could lead to confusion. However, we definitely want to have both integration health entries for this.

I think we have to come up with a way here to handle this generically. It'd be best to add a TODO here and handle it in a separate PR. It will also affect us once we add other integrations for which health status already exists (i.e. image integrations).

Reply:
That's an issue. I'll address it in follow-up by adding prefix to declarative configurations integration health ID.

Reply:
- https://github.com/stackrox/stackrox/pull/6198